File: rust/src/engine/repeat_keyed.rs
//! `repeat=` built in memory the way the streaming engine builds it.
//!
//! A repeating column has two plans, not one. How MANY values a row keeps is an exact quota
//! over the run — permuted by `#replen`, so a row's length follows from its own position and
//! never from a running total over its predecessors. What those values ARE then depends on the
//! generator: a list is laid out over the whole slot space and read at the row's slots, while
//! anything drawn takes one seekable sub-stream per element, `#e0`, `#e1`, and so on.
//!
//! Both halves are keyed by `(seed, stream_id)` and mirror the reference's `repeat-keyed.ts`.
//! The older sequential builder in `generators/repeat.rs` stays for the cases with nothing to
//! key by — an inline generator inside a pack body.

use super::EngineResult;
use crate::generators::repeat;
use crate::prng::{permute, seekable, Sfc32};
use crate::stats::hamilton;

use super::per_row::Stream;

/// How many values each position keeps, and where in the slot space they start.
fn length_plan(spec: &repeat::Spec, count: usize, stream: &Stream) -> (repeat::Plan, i32) {
    let mut pct = crate::prng::create(&format!("{}|{}|replen", stream.seed, stream.id));
    let counts = hamilton::counts_per_value(count as i32, &repeat::length_percents(spec), &mut pct);
    let key = permute::key(&stream.seed, &format!("{}#replen", stream.id));
    (repeat::plan(spec, &counts), key)
}

/// A repeating column of DRAWN values.
///
/// Element k of a row comes off the row's own `#e{k}` stream, so the row still resolves alone —
/// which is also what lets a worker render a range of rows without seeing the rest.
pub fn build_draws(
    spec: &repeat::Spec,
    count: usize,
    stream: &Stream,
    mut one_element: impl FnMut(usize, &mut Sfc32, &mut [bool]) -> EngineResult<String>,
    mut flag_text_out: Option<&mut Vec<String>>,
) -> EngineResult<Vec<String>> {
    let (plan, key) = length_plan(spec, count, stream);
    let mut out = Vec::with_capacity(count);
    for i in 0..count {
        let row = stream.row_at(i);
        let keep = plan.length_at(permute::apply(i as i32, count as i32, key) as usize);
        let mut parts = Vec::with_capacity(keep);
        let mut marks = Vec::with_capacity(keep);
        for k in 0..keep {
            let mut element_prng =
                seekable::generator(&stream.seed, &format!("{}#e{k}", stream.id), row as i32);
            let mut flag = [false];
            parts.push(one_element(k, &mut element_prng, &mut flag)?);
            marks.push(if flag[0] { "true" } else { "false" }.to_string());
        }
        out.push(repeat::join(&parts, spec)?);
        // A parallel list of true/false, never a running total — accumulating it would mean
        // nothing — so it joins with the separator alone.
        if let Some(flags) = flag_text_out.as_deref_mut() {
            flags.push(marks.join(&spec.separator));
        }
    }
    Ok(out)
}

/// A repeating column of LISTED values.
///
/// The slot space covers every element of every row at once, laid out exactly and permuted; a
/// row reads the slots its length plan gave it.
pub fn build_layout(
    spec: &repeat::Spec,
    values: &[String],
    percents: &[f64],
    count: usize,
    stream: &Stream,
    mut modify: impl FnMut(usize, String, usize) -> String,
) -> EngineResult<Vec<String>> {
    let (plan, len_key) = length_plan(spec, count, stream);
    let slots = plan.total_slots;
    let mut pct = crate::prng::create(&format!("{}|{}|pct", stream.seed, stream.id));
    let counts = hamilton::counts_per_value(slots as i32, percents, &mut pct);
    let key = permute::key(&stream.seed, &stream.id);

    let mut cum_hi = Vec::with_capacity(counts.len());
    let mut acc = 0;
    for c in &counts {
        acc += c;
        cum_hi.push(acc);
    }
    let value_for_slot = |slot: i32| -> String {
        let (mut lo, mut hi) = (0usize, cum_hi.len().saturating_sub(1));
        while lo < hi {
            let mid = (lo + hi) / 2;
            if slot < cum_hi[mid] {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        values.get(lo).cloned().unwrap_or_default()
    };

    let mut out = Vec::with_capacity(count);
    for i in 0..count {
        let p = permute::apply(i as i32, count as i32, len_key) as usize;
        let row = stream.row_at(i);
        let start = plan.slot_start_at(p);
        let keep = plan.length_at(p);
        let mut parts = Vec::with_capacity(keep);
        for k in 0..keep {
            let raw = value_for_slot(permute::apply((start + k) as i32, slots as i32, key));
            parts.push(modify(row, raw, k));
        }
        out.push(repeat::join(&parts, spec)?);
    }
    Ok(out)
}

/// The `anomaly=`/`missing=` draw for one element of a repeating LISTED column.
///
/// One draw per element, pulled a whole row at a time — the budget is the row's maximum
/// length, so which uniform element k gets does not depend on how long its row turned out.
pub fn element_uniforms<'a>(
    stream: &'a Stream,
    purpose: &str,
    budget: usize,
) -> impl FnMut(usize, usize) -> f64 + 'a {
    let id = format!("{}{purpose}", stream.id);
    let mut cached: Option<(usize, Vec<f64>)> = None;
    move |row: usize, k: usize| -> f64 {
        if cached.as_ref().map(|(r, _)| *r) != Some(row) {
            cached = Some((
                row,
                seekable::uniforms(&stream.seed, &id, row as i32, budget),
            ));
        }
        cached
            .as_ref()
            .and_then(|(_, drawn)| drawn.get(k).copied())
            .unwrap_or(1.0)
    }
}

File: java/src/main/java/io/github/nickliapin/tdc/engine/RepeatKeyed.java
package io.github.nickliapin.tdc.engine;

import io.github.nickliapin.tdc.distribution.Hamilton;
import io.github.nickliapin.tdc.generators.Repeat;
import io.github.nickliapin.tdc.prng.Permute;
import io.github.nickliapin.tdc.prng.Prng;
import io.github.nickliapin.tdc.prng.Seekable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * {@code repeat=} built in memory the way the streaming engine builds it.
 *
 * <p>A repeating column has two plans, not one. How MANY values a row keeps is an exact quota over
 * the run — permuted by {@code #replen}, so a row's length follows from its own position and never
 * from a running total over its predecessors. What those values ARE then depends on the generator:
 * a list is laid out over the whole slot space and read at the row's slots, while anything drawn
 * takes one seekable sub-stream per element, {@code #e0}, {@code #e1}, and so on.
 *
 * <p>Both halves are keyed by {@code (seed, streamId)} and mirror the reference's {@code
 * repeat-keyed.ts}. The older sequential builder in {@link Repeat#build} stays for the cases with
 * nothing to key by — an inline generator inside a pack body.
 */
final class RepeatKeyed {

  private RepeatKeyed() {}

  /** One element of one row: its own generator, and a one-slot flag for {@code anomaly=}. */
  interface Element {
    String build(int k, Prng.Sfc32 prng, boolean[] flag);
  }

  /** One element of a LISTED row: the row it belongs to, its raw value, and its index in the row. */
  interface Modifier {
    String apply(int row, String value, int k);
  }

  /**
   * A repeating column of DRAWN values.
   *
   * <p>Element k of a row comes off the row's own {@code #e{k}} stream, so the row still resolves
   * alone — which is also what lets a worker render a range of rows without seeing the rest.
   */
  static List<String> buildDraws(
      Repeat.Spec spec,
      int count,
      PerRow.Stream stream,
      String genType,
      Element element,
      List<String> flagTextOut) {
    Repeat.Plan plan = lengthPlan(spec, count, stream);
    int key = Permute.key(stream.seed(), stream.id() + "#replen");
    List<String> out = new ArrayList<>(count);
    for (int i = 0; i < count; i++) {
      int row = stream.rowAt(i);
      int keep = plan.lengthAt(Permute.permute(i, count, key));
      List<String> parts = new ArrayList<>(keep);
      List<String> marks = new ArrayList<>(keep);
      for (int k = 0; k < keep; k++) {
        boolean[] flag = new boolean[1];
        final int at = k;
        final int atRow = row;
        // A drawn generator has no pool to draw down, so `distinct` is rejection sampling on
        // fresh sub-streams — the same ids the reference uses, so the two agree value for value.
        java.util.function.Function<String, String> drawAt =
            suffix ->
                element.build(
                    at,
                    Seekable.generator(stream.seed(), stream.id() + "#e" + at + suffix, atRow),
                    flag);
        parts.add(
            spec.distinct()
                ? Repeat.redrawUntilFresh(parts, genType, drawAt)
                : drawAt.apply(""));
        marks.add(flag[0] ? "true" : "false");
      }
      out.add(Repeat.join(parts, spec));
      // A parallel list of true/false, never a running total — accumulating it would mean nothing
      // — so it joins with the separator alone.
      if (flagTextOut != null) {
        flagTextOut.add(String.join(spec.separator(), marks));
      }
    }
    return out;
  }

  /**
   * A repeating column of LISTED values.
   *
   * <p>The slot space covers every element of every row at once, laid out exactly and permuted; a
   * row reads the slots its length plan gave it.
   */
  static List<String> buildLayout(
      Repeat.Spec spec,
      List<String> values,
      double[] percents,
      int count,
      PerRow.Stream stream,
      Modifier modify) {
    Repeat.Plan plan = lengthPlan(spec, count, stream);
    int lengthKey = Permute.key(stream.seed(), stream.id() + "#replen");
    int slots = plan.totalSlots();
    int[] counts =
        Hamilton.countsPerValue(
            slots, percents, Prng.create(stream.seed() + "|" + stream.id() + "|pct"));
    int key = Permute.key(stream.seed(), stream.id());

    int[] cumHi = new int[counts.length];
    int acc = 0;
    for (int i = 0; i < counts.length; i++) {
      acc += counts[i];
      cumHi[i] = acc;
    }

    List<String> out = new ArrayList<>(count);
    for (int i = 0; i < count; i++) {
      int p = Permute.permute(i, count, lengthKey);
      int row = stream.rowAt(i);
      int start = plan.slotStartAt(p);
      int keep = plan.lengthAt(p);
      List<String> parts = new ArrayList<>(keep);
      // `distinct` leaves the whole-run layout behind: a row that must not repeat itself has to
      // CHOOSE from the pool, and a choice cannot be read off a pre-laid-out slot. One uniform
      // per pick off the row's own stream, budgeted at the maximum length, so the row still
      // resolves alone.
      if (spec.distinct()) {
        double[] draws = Seekable.uniforms(stream.seed(), stream.id() + "#dist", row, spec.max());
        int[] at = {0};
        List<String> picked =
            Repeat.drawDistinct(
                values,
                percents,
                keep,
                () -> at[0] < draws.length ? draws[at[0]++] : 1.0,
                "the value list");
        for (int k = 0; k < picked.size(); k++) {
          String raw = picked.get(k);
          parts.add(modify == null ? raw : modify.apply(row, raw, k));
        }
      } else {
        for (int k = 0; k < keep; k++) {
          String raw = valueForSlot(values, cumHi, Permute.permute(start + k, slots, key));
          parts.add(modify == null ? raw : modify.apply(row, raw, k));
        }
      }
      out.add(Repeat.join(parts, spec));
    }
    return out;
  }

  /**
   * The {@code anomaly=}/{@code missing=} draw for one element of a repeating LISTED column.
   *
   * <p>One draw per element, pulled a whole row at a time — the budget is the row's maximum length,
   * so which uniform element k gets does not depend on how long its row turned out.
   */
  static ElementDraw elementUniforms(PerRow.Stream stream, String purpose, int budget) {
    String id = stream.id() + purpose;
    Map<Integer, double[]> cache = new HashMap<>(1);
    return (row, k) -> {
      double[] drawn = cache.get(row);
      if (drawn == null) {
        drawn = Seekable.uniforms(stream.seed(), id, row, budget);
        cache.clear();
        cache.put(row, drawn);
      }
      return k < drawn.length ? drawn[k] : 1.0;
    };
  }

  /** The uniform element {@code k} of {@code row} gets. */
  interface ElementDraw {
    double at(int row, int k);
  }

  /** How many values each position keeps, and where in the slot space they start. */
  private static Repeat.Plan lengthPlan(Repeat.Spec spec, int count, PerRow.Stream stream) {
    int[] counts =
        Hamilton.countsPerValue(
            count,
            Repeat.lengthPercents(spec),
            Prng.create(stream.seed() + "|" + stream.id() + "|replen"));
    return Repeat.plan(spec, count, counts);
  }

  private static String valueForSlot(List<String> values, int[] cumHi, int slot) {
    int lo = 0;
    int hi = cumHi.length - 1;
    while (lo < hi) {
      int mid = (lo + hi) / 2;
      if (slot < cumHi[mid]) {
        hi = mid;
      } else {
        lo = mid + 1;
      }
    }
    return lo < values.size() ? values.get(lo) : "";
  }
}
